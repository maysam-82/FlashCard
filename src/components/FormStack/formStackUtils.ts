import { ICard } from '../../types/card';

export const cardVerificationResult = (cards: ICard[]): boolean => {
	if (cards.length > 0) {
		return !!cards.find((card) => !card.prompt || !card.answer);
	}
	return false;
};
