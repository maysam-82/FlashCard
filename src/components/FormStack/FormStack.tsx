import React from 'react';
import {
	Form,
	FormControl,
	Button,
	FormGroup,
	FormLabel,
} from 'react-bootstrap';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';
import { ICard } from '../../types/card';
import { addStack } from '../../redux/actions';
import { cardVerificationResult } from './formStackUtils';

export interface IState {
	id: number;
	title: string;
	cards: ICard[];
}

export interface IProps {
	addStack: typeof addStack;
}

export class FormStack extends React.Component<
	IProps & RouteComponentProps,
	IState
> {
	constructor(props: IProps & RouteComponentProps) {
		super(props);

		this.state = {
			title: '',
			cards: [],
			id: 0,
		};
	}

	addCardHandler = () => {
		if (!cardVerificationResult(this.state.cards)) {
			this.setState((prevState) => {
				const newCard: ICard = {
					id: prevState.cards.length,
					prompt: '',
					answer: '',
				};
				return { cards: [...prevState.cards, newCard] };
			});
		}
	};

	updateCardHandler = (
		event: React.ChangeEvent<HTMLInputElement>,
		id: number,
		key: keyof ICard = 'prompt'
	) => {
		const { value } = event.target;
		this.setState((prevState) => {
			const { cards } = prevState;
			// https://github.com/microsoft/TypeScript/issues/31663#issuecomment-519563197
			(cards[id][key] as any) = value;
			return { cards: [...cards] };
		});
	};
	renderCards = () => {
		const { cards } = this.state;
		return this.state.cards.map(({ id }) => {
			return (
				<FormGroup
					key={id}
					className="border-bottom p-2 mb-2 d-flex align-items-center"
				>
					<FormLabel className="m-0">Prompt:</FormLabel>
					<FormControl
						onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
							this.updateCardHandler(event, id, 'prompt')
						}
						value={cards[id].prompt}
						className="ml-2 mr-4"
					/>
					<FormLabel className="m-0 ml-2">Answer:</FormLabel>
					<FormControl
						onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
							this.updateCardHandler(event, id, 'answer')
						}
						value={cards[id].answer}
						className="ml-2"
					/>
				</FormGroup>
			);
		});
	};

	addStackHandler = () => {
		const { title, cards } = this.state;
		if (
			title &&
			cards.length > 0 &&
			!cardVerificationResult(this.state.cards)
		) {
			const id = Date.now();
			this.setState({ id }, () => {
				this.props.addStack(this.state);
				this.props.history.push('/');
			});
		}
	};

	render() {
		const { title } = this.state;
		return (
			<div className="form-stack-container">
				<Form inline className="p-2 border-bottom mb-2">
					<FormGroup className="mb-2">
						<FormLabel className="mr-2">Title:</FormLabel>
						<FormControl
							placeholder="Enter title of card"
							value={title}
							onChange={(event) => this.setState({ title: event.target.value })}
						/>
					</FormGroup>
				</Form>
				<div className="cards-list">{this.renderCards()}</div>
				<Button onClick={this.addCardHandler} className="ml-2">
					Add Card
				</Button>
				<Button onClick={this.addStackHandler} className="ml-2 bg-success">
					Save And Add Stack
				</Button>
			</div>
		);
	}
}

export default connect(null, { addStack })(FormStack);
