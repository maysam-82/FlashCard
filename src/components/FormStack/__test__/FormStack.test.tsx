import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { RouteComponentProps, match } from 'react-router-dom';
import * as H from 'history';
import { FormStack, IProps, IState } from '../FormStack';

// Shallow test with ReactRouter and white box test
const props: IProps & RouteComponentProps = {
	addStack: jest.fn(),
	history: {
		push: (path: '') => {},
	} as H.History,
	location: {} as H.Location,
	match: {} as match,
};

describe('FormStack component', () => {
	let component: ShallowWrapper<IProps, IState, FormStack>;
	component = shallow<FormStack, IProps, IState>(<FormStack {...props} />);
	it('should render FormStack', () => {
		expect(component).toMatchSnapshot();
	});
	it('should render `Form` component', () => {
		expect(component.find('Form').exists()).toBeTruthy();
	});
	it('should render `Add Card` `Button`', () => {
		expect(component.find('Button').at(0).props().children).toEqual('Add Card');
	});
	it('should render `Save And Add Stack` `Button` ', () => {
		expect(component.find('Button').at(1).props().children).toEqual(
			'Save And Add Stack'
		);
	});
	describe('When user change value in `FormControl`', () => {
		beforeEach(() => {
			component
				.find('FormControl')
				.simulate('change', { target: { value: 'test title stack' } });
		});
		afterEach(() => {
			component
				.find('FormControl')
				.simulate('change', { target: { value: '' } });
		});
		it('should update `title` in the `state`', () => {
			expect(component.state().title).toEqual('test title stack');
		});
	});
	describe('When user clicks on `Add Card` `Button`', () => {
		beforeEach(() => {
			component.find('Button').at(0).simulate('click');
		});
		afterEach(() => {
			component.setState({ cards: [] });
		});
		it('should add one new card to `cards` in `state`', () => {
			expect(component.state().cards.length).toEqual(1);
		});
	});
	describe('When user clicks on `Save And Add Stack` `Button` When `state.title` is empty', () => {
		beforeEach(() => {
			component.find('Button').at(1).simulate('click');
		});
		it('should not update state', () => {
			expect(component.state()).toEqual({ title: '', cards: [], id: 0 });
		});
		describe('When user clicks on `Save And Add Stack` `Button` When `state.title` is not empty', () => {
			beforeEach(() => {
				component.find('Button').at(1).simulate('click');
				component
					.find('FormControl')
					.simulate('change', { target: { value: 'test title stack' } });
			});
			it('should not call `addStack`', () => {
				expect(component.instance().props.addStack).toHaveBeenCalledTimes(0);
			});
			describe('When there is an object inside `cards` in `state` but its contains is not verified', () => {
				beforeEach(() => {
					component.find('Button').at(0).simulate('click');
				});
				it('should not call `addStack`', () => {
					expect(component.instance().props.addStack).toHaveBeenCalledTimes(0);
				});
				describe('When `cards` contains verified objects', () => {
					beforeEach(() => {
						const mockEventChangePrompt: any = {
							target: { value: 'test prompt value' },
						};
						const mockEventChangeAnswer: any = {
							target: { value: 'test answer value' },
						};
						component
							.instance()
							.updateCardHandler(mockEventChangePrompt, 0, 'prompt');
						component
							.instance()
							.updateCardHandler(mockEventChangeAnswer, 0, 'answer');
					});
					it('should create a unique `id` in the state', () => {
						component.find('Button').at(1).simulate('click');
						expect(component.state().id).not.toBe(0);
					});
				});
			});
		});
	});
});
