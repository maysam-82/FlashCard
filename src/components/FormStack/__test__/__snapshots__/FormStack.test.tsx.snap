// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormStack component should render FormStack 1`] = `
<div
  className="form-stack-container"
>
  <Form
    className="p-2 border-bottom mb-2"
    inline={true}
  >
    <FormGroup
      className="mb-2"
    >
      <FormLabel
        className="mr-2"
        column={false}
        srOnly={false}
      >
        Title:
      </FormLabel>
      <FormControl
        onChange={[Function]}
        placeholder="Enter title of card"
        value=""
      />
    </FormGroup>
  </Form>
  <div
    className="cards-list"
  />
  <Button
    active={false}
    className="ml-2"
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Add Card
  </Button>
  <Button
    active={false}
    className="ml-2 bg-success"
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Save And Add Stack
  </Button>
</div>
`;
